### PR TITLE
Plugins: Change managed plugins installation call

### PR DESCRIPTION
**Pull request by oshirohugo**
**What is this feature?**

Change manage plugins installation call:
* Add feature toggle check to use or not the old installation
* Allow manage plugins to call backend installer (there will be a proper implementation plugin installer in managed instances)
* Add a modal to show that the plugin won't be available right away

**Why do we need this feature?**

This feature is needed to improve managed plugin installation https://github.com/grafana/grafana-plugins-platform-team/issues/21

**Who is this feature for?**

Plugin catalog users in Grafana managed instances.

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana-plugins-platform-team/issues/28

**Special notes for your reviewer:**

https://github.com/grafana/grafana/assets/5558280/56506965-8ba5-4597-9468-69d6bcb3a3c4

To be able to test the full solution you need to set the following:
```ini
[plugins]
plugin_admin_external_manage_enabled = true
install_token = "<token-with-scopes-stack-plugin-read-write-delete>"

[feature_toggles]
enable = managedPluginsInstall
managedPluginsInstall = true

[environment]
stack_id = "<stack-id>"
```

you can use your personal stack-id and a personal policy/token.

the enterprise PR:
https://github.com/grafana/grafana-enterprise/pull/6009

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by sunker** [comment]
Was trying to test this locally, but couldn't get the modal to be displayed. Am I missing something in my ini file?

```yaml
[feature_toggles]
managedPluginsInstall=true

[plugins]
pluginAdminExternalManageEnabled=true
````

**Comment by oshirohugo** [comment]
@sunker 
Thanks for checking the PR.
I think you need to use
```ini
[plugin]
plugin_admin_external_manage_enabled = true
```
instead of
```ini
[plugins]
pluginAdminExternalManageEnabled=true
```

**Comment by andresmgot** [comment]
@sympatheticmoose may want to take a look to the wording / UX.

In my local instance, the button gets disabled with a message "Installing..." until it installs it and then the modal is opened. I suspect the UX will be different in cloud. How would that work there?

**Comment by oshirohugo** [comment]
> In my local instance, the button gets disabled with a message "Installing..." until it installs it and then the modal is opened. I suspect the UX will be different in cloud. How would that work there?

In cloud most of things will be similar:
"The button gets disabled with a message "Installing..." until it installs it and then the modal is opened." In cloud the toaster saying that the plugin could not be found also will be popped, since the instance doesn't have the plugin yet.

Everything else should be the same. The user just won't be able to use the plugin right away.

**Comment by sunker** [comment]
Looks good to me! Possibly a bit unintuitive that the modal pops up after a few seconds and not immediately.

**Comment by andresmgot** [comment]
Cool, thanks for the video! I _think_ it would be better to modify the toaster text in this case. Now there is information about the installation poping up in two different places, making it a bit confusing. What I would do is changing the text of the toaster from: "Installed < plugin >" to "< plugin > submitted for installation. It can take several minutes to be available" (or something along those lines).

Also, what happens if you refresh the page and the plugin is not fully installed yet?

**Comment by oshirohugo** [comment]
> What I would do is changing the text of the toaster from: "Installed < plugin >" to "< plugin > submitted for installation. It can take several minutes to be available" (or something along those lines).

I'll change it

> Also, what happens if you refresh the page and the plugin is not fully installed yet?

You're right, there is indeed a problem here. If you refresh the page the install button will be available again. To avoid this I will have to create a new endpoint in grafana backend to be able to get the `/instances/{instance-id}/plugins` from GCOM and check if this is installed. This needs to be done in the backend because it uses the token. (if you see a better way of doing it let me know)
I'll create a follow up PR for that.

**Comment by andresmgot** [comment]
> You're right, there is indeed a problem here. If you refresh the page the install button will be available again. To avoid this I will have to create a new endpoint in grafana backend to be able to get the `/instances/{instance-id}/plugins` from GCOM and check if this is installed. This needs to be done in the backend because it uses the token. (if you see a better way of doing it let me know)
> I'll create a follow up PR for that.

sounds good :+1:

**Review by sympatheticmoose — COMMENTED** [review]
Thanks for the work on this 🙌 

Could you help me understand this better:
> In cloud the toaster saying that the plugin could not be found also will be popped, since the instance doesn't have the plugin yet. 

I don't see that in the video and not sure what its referring to

---

Generally I'd like to understand more about why we need both the modal and the toast/AppEvent:
- whether changing the wording/type on the event would be sufficient
- (or) impact of suppressing that event from being fired.
- I can offer some suggestions on wordings for both once I understand the above a bit more

In the video it goes straight to uninstall, is the video from a local instance? If cloud, wondering what happens if user requests uninstall before the install has completed. (I'm assuming its local, and if Cloud we don't have the uninstall logic yet)

**Comment by oshirohugo** [comment]
Thanks for your review @sympatheticmoose 
> I don't see that in the video and not sure what its referring to
This doesn't appear in the video, but it will appear in a cloud instance since after installation we call the backend to check the the plugin is indeed installed, the backend checks the file system for that and as it won't be there right away, it will show a toaster saying it's not found. This is a check that already exists in grafana and that's how grafana checks if plugins are installed.
I hope this gives you a better understanding.

> In the video it goes straight to uninstall, is the video from a local instance? If cloud, wondering what happens if user requests uninstall before the install has completed. (I'm assuming its local, and if Cloud we don't have the uninstall logic yet)

The video is indeed a local instance, and it was made only to show the intended behavior on cloud.
In cloud, from what I could test, if you click uninstall button before the new instance is created, the process will be overwritten, and it will seem like you never did the installation.
Right now the same behavior can be tested with the grafana.com installation, if you install a plugin in your instance use grafana.com and uninstall it right away also using grafana.com, in your instance usage it will seem like nothing happened. What we are doing is just cutting the part where the user have to leave the catalog to go to grafana.com to do all this.

**Comment by andresmgot** [comment]
> Generally I'd like to understand more about why we need both the modal and the toast/AppEvent:
> 
> * whether changing the wording/type on the event would be sufficient

From what I understood, the goal was to change the wording in the toast and remove the modal. Did I get that right or you were still planning on keeping the modal @oshirohugo ?

**Comment by oshirohugo** [comment]
@andresmgot you're right. This was confusing in my answer.

@sympatheticmoose as Andres suggested we will remove the modal and just let the toaster with the info.
I'll update the video soon, so you can have a better idea. sorry for the confusion.

**Comment by sympatheticmoose** [comment]
>This doesn't appear in the video, but it will appear in a cloud instance since after installation we call the backend to check the the plugin is indeed installed, the backend checks the file system for that and as it won't be there right away, it will show a toaster saying it's not found. This is a check that already exists in grafana and that's how grafana checks if plugins are installed.
I hope this gives you a better understanding.

It does - thank you 🙌 and we don't have a way of preventing that being displayed to the user?

**Comment by oshirohugo** [comment]
@sympatheticmoose @andresmgot I've updated the video with the message in the toaster, please give me your thoughts

**Comment by andresmgot** [comment]
> @sympatheticmoose @andresmgot I've updated the video with the message in the toaster, please give me your thoughts

I still think it's a bit confusing because:

  - The existing (green) toaster says that the plugin has been successfully installed, which is not true (yet).
  - The new toaster says there is an "Error", which contradicts the other one.

IMO, I would change the content of the existing toaster (green) and just say something like "Plugin <ID> marked for installation. Note that it can take some minutes for the plugin to be fully installed." (or something on that line).

File: public/app/features/plugins/admin/components/PluginActions.tsx
Comment (andresmgot) [inline public/app/features/plugins/admin/components/PluginActions.tsx:40]: I don't think you need the extra wrapping?

**Review by andresmgot — APPROVED** [review]
Code LGTM, maybe you want to have additional frontend eyes here.

Regarding the wording I'll let @sympatheticmoose to have a final review.

**Comment by oshirohugo** [comment]
@sunker I've done some significant changes to follow the CR suggestion, could you please take another look?

@sympatheticmoose I've updated the video with the last changes.

File: public/app/features/plugins/admin/components/InstallControls/InstallControlsButton.tsx
Comment (sympatheticmoose) [inline public/app/features/plugins/admin/components/InstallControls/InstallControlsButton.tsx:73]: When we took this to vote in slack the winner was `Install requested, this may take a few minutes.` 🙂

Comment (oshirohugo) [reply public/app/features/plugins/admin/components/InstallControls/InstallControlsButton.tsx]: @sympatheticmoose I've updated the message and the video, please take a look again!

**Review by sympatheticmoose — APPROVED** [review]
Looking good 🚀

**Review by sunker — APPROVED** [review]
Looks good to me! 

Feels a little off to `alertSuccess` since we at that time don't know if plugin installation will work or not. Wish there was an `alertInfo` app event...anyway, not a big deal. Great work on this!

**Pull request merged**
